pkg/ensurance/cm/advanced_cpu_manager.go

```diff
@@ -0,0 +1,351 @@
+package cm
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	pb "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/kubelet/cm/containermap"
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpumanager/state"
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpumanager/topology"
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
+
+	"github.com/gocrane/crane/pkg/ensurance/collector/cadvisor"
+	cruntime "github.com/gocrane/crane/pkg/ensurance/runtime"
+)
+
+const (
+	// cpuManagerStateFileName is the file name where cpu manager stores its state
+	cpuManagerStateFileName = "cpu_manager_state"
+	craneCpusetPolicyName   = "crane"
+	// stateFilePath holds the directory where the state file for checkpoints is held.
+	stateFilePath        = "/rootvar/run/crane"
+	kubeletStateFilePath = "/kubelet/"
+	// cpusetReconcilePeriod  is the duration between calls to reconcileState.
+	cpusetReconcilePeriod = 5 * time.Second
+	// intervalRetryAddContainer is the interval between try add container
+	intervalRetryAddContainer = 200 * time.Millisecond
+	// timeoutRetryAddContainer is the timeout for adding container
+	timeoutRetryAddContainer = 2 * time.Second
+)
+
+type AdvancedCpuManager struct {
+	isStarted bool
+
+	sync.RWMutex
+	policy Policy
+
+	podLister corelisters.PodLister
+	podSynced cache.InformerSynced
+
+	runtimeClient pb.RuntimeServiceClient
+	runtimeConn   *grpc.ClientConn
+
+	// reconcilePeriod is the duration between calls to reconcileState.
+	reconcilePeriod time.Duration
+
+	// state allows pluggable CPU assignment policies while sharing a common
+	// representation of state for the system to inspect and reconcile.
+	state state.State
+
+	// stateFileDirectory holds the directory where the state file for checkpoints is held.
+	stateFileDirectory string
+}
+
+func NewAdvancedCpuManager(client clientset.Interface, nodeName string, podInformer coreinformers.PodInformer,
+	nodeInformer coreinformers.NodeInformer, runtimeEndpoint string) *AdvancedCpuManager {
+	cadvisorManager, err := cadvisor.NewCadvisorManager()
```

Comment:
We think is better to use the one in stateCollector,  and no need to init another one. Use `stateCollector.GetCollectors()` in https://github.com/gocrane/crane/pull/196/files.

Reply:
ok

Reply:
rely on more than one nep open now..improve in 0.4.0